### DISCUSSION TITLE
[question] why type job.id is optional? isn's it fallback to default value when not provided by user.

### DISCUSSION THREAD

**Discussion by michael-land** [taskforcesh/bullmq#844] [Q&A]
https://github.com/taskforcesh/bullmq/blob/master/src/classes/job.ts#L142

**Comment by manast** [comment]
https://docs.bullmq.io/guide/jobs/job-ids

**Reply by michael-land** [reply]
@manast 

To make it clear, I think the job (instance returned by `queue.add()`) id type should be `string` instead of `string|undefined`.

===

Currently, The job#id is `string|undefined` 

![image](https://user-images.githubusercontent.com/33362998/139362009-9de0d807-0055-4c83-9c36-79d9172649b4.png)

> By default job ids are generated automatically as an increasing counter

according to the doc, job's id property should be non-nullable.

should we change the signature to something like this? it take an optional custom id, and change the job's id type to `string`

```ts
export class Job<DataType = any, ReturnType = any, NameType extends string = string> {
  id: string;

  constructor(
    private queue: MinimalQueue,
    /**
     * The name of the Job
     */
    public name: NameType,

    /**
     * The payload for this job.
     */
    public data: DataType,

    /**
     * The options object for this job.
     */
    public opts: JobsOptions = {},
    public customId?: string
  ) {
    // this will initialized later if user does not provide one
    this.id = customId as string 
  }
}

```

**Reply by manast** [reply]
yeah, you are right, the job.id should not be optional.

**Reply by craig-feldman** [reply]
@manast just resurrecting this discussion. Could the typings please be updated to reflect the fact that a persisted job contains a defined id?
